Blog on types of CrossGuard policies you can write (#2179)
pulumi/docs: content/blog/writing-crossguard-policies/index.md

---

> title: "Writing CrossGuard Policies"

Reviewer: Nit: Can we jazz this up any? "Enforcing different kinds of policies for cloud resources"?  "Writing X policies" sounds super generic. While name-dropping "CrossGuard" is nice, it might not be catchy.

date: 2019-12-19
meta_desc: "A look at the different types of policies that can be written for Pulumi CrossGuard."
meta_image: crossguard.png
authors:
    - justin-vanpatten
tags:
    - crossguard
    - policy-as-code
---

> We recently announced [a new policy as code solution, CrossGuard]({{< relref "/blog/announcing-crossguard-preview" >}}) that validates policies at deployment time. Policies are expressed as code and are used to prevent the creation of out-of-compliance resources. Since policies are written using full-blown programming languages, it's possible to do interesting things such as [combining IAM Access Analyzer and Pulumi CrossGuard]({{< relref "/blog/aws-iam-access-analyzer-and-crossguard" >}}). In this post, we'll take a closer look at the different types of policies that can be written.

Reviewer: Seems like we should add a "what's in it for you" type sentence in the middle of the paragraph, e.g.:

... used to prevent the creation of out-of-compliance resources. [This allows an organization to prevent entire classes of security and reliability defects to ensure infrastructure is following best practices.] Since policies are written...

<!--more-->

## Authoring Policies

> A policy contains specific logic you want to enforce. For example, you may want to prevent the creation of public, world-readable storage objects (e.g. on AWS S3, Azure BlobStore, etc.) or prevent the creation of a virtual machine without the proper security groups in-place. Policies are run during `pulumi preview` and `pulumi update`, ensuring that cloud resource definitions are in compliance.

Reviewer: My understanding is that "security groups" is an AWS-specific thing. Perhaps use "proper firewalls" or "secure networking rules" instead, if that is more common?

Policies are written as validation functions that are evaluated against all resources in your Pulumi stack. If the validation function calls `reportViolation`, the associated resource will be considered in violation of the policy.

> There are two types of policies:

Reviewer: What do you think about making this a table instead, does the markup make it easier to read?
---
There are two types of policies, each enforcing a different class of restriction.

| Policy Kind | Description |
| --- | --- |
| `ResourceValidationPolicy` | ... |
| ... | ... |

Also, can we link `ResourceValidationPolicy` to the H2 in the document or documentation?

    name: "elb-logging-enabled",
    description: "Checks whether the Application Load Balancers and the Classic Load Balancers have logging enabled.",
    enforcementLevel: "mandatory",
    validateResource: [
        validateTypedResource(aws.elasticloadbalancing.LoadBalancer, (loadBalancer, args, reportViolation) => {
            if (loadBalancer.accessLogs === undefined || !loadBalancer.accessLogs.enabled) {
                reportViolation("Elastic Load Balancer must have access logs enabled.");
            }
        }),
        validateTypedResource(aws.elasticloadbalancingv2.LoadBalancer, (loadBalancer, args, reportViolation) => {
            if (loadBalancer.accessLogs === undefined || !loadBalancer.accessLogs.enabled) {
                reportViolation("Elastic Load Balancer must have access logs enabled.");
            }
        }),
        validateTypedResource(aws.applicationloadbalancing.LoadBalancer, (loadBalancer, args, reportViolation) => {
            if (loadBalancer.accessLogs === undefined || !loadBalancer.accessLogs.enabled) {
                reportViolation("Application Load Balancer must have access logs enabled.");
            }
        }),
    ],
};
```

> 

Reviewer: We should add a quick paragraph about the drawbacks of a resource validation policy too. For example, since it is ran before the resource is created, you cannot verify things like the assigned name or properties the "live" state of the resource has.

So for example, some types of policies cannot be expressed in a resource validation policy. But the good news is that there are other types of policies you can write...

## Stack Validation Policies

> Policies of `StackValidationPolicy` are run against the stack as a whole. These policies are run _after_ all stack resources are registered and thus do not block an out-of-compliance resource from being created or modified, but do fail the preview or update. To avoid creating out-of-compliance resources, we recommend running a preview command before an update. However, there are cases where a policy requires a resource to exist in order to perform its validation, such as inspecting the resource's output properties that are only available to already provisioned resources (e.g. checking if a resource's `id` is referenced by another resource). In such cases, the policy can skip reporting violations during previews and only report violations during updates when the information it needs is available. Thus, the preview will succeed and the resources will be created or modified during the update, but the overall update operation will fail if any resources are in violation.

Reviewer 2: ```suggestion
Policies of `StackValidationPolicy` run against the whole stack. These policies are run _after_ all stack resources are registered and thus do not block an out-of-compliance resource from being created or modified, but do fail the preview or update. To avoid creating out-of-compliance resources, we recommend running a preview command before an update. 

However, there are cases where a resource must exist for a policy to validate it, such as inspecting the resource's output properties that are only available to already provisioned resources (e.g., checking if a resource's `id` is referenced by another resource). In such cases, the policy can skip reporting violations during previews and only report violations during updates when the information it needs is available. Thus, the preview can succeed. and the resources are created or modified during the update. The overall update operation fails if any resources are in violation.
```

> 

Reviewer: Maybe also mention that this allows you to write policies that require inspecting _multiple_ resources at once. For example, that a virtual machine and its persistent hard drive are in the same stack, and have congruent settings. (or something).

> A stack validation function is passed `args` with information about all of the resources in the stack (such as each resource's type, _output_ properties, name, and URN) and a `reportViolation` callback that can be called to report a policy violation. `reportViolation` requires passing a `message` and also supports an optional `urn` parameter, to specify the URN of the resource in violation (otherwise violation is associated with the stack itself).

Reviewer: Nit: The `urn` of a resource is probably unknown to the reader... maybe it's just confusing and we should cut it? Or alternatively, link to somewhere in the docs where we explain what it is?

> A stack validation function is passed `args` with information about all of the resources in the stack (such as each resource's type, _output_ properties, name, and URN) and a `reportViolation` callback that can be called to report a policy violation. `reportViolation` requires passing a `message` and also supports an optional `urn` parameter, to specify the URN of the resource in violation (otherwise violation is associated with the stack itself).

Reviewer 2: ```suggestion
A stack validation function is passed `args` with information about all of the resources in the stack (such as each resource's type, _output_ properties, name, and URN) and a `reportViolation` callback for reporting a policy violation. A `reportViolation` requires passing a `message` and also supports an optional `urn` parameter for the URN of the resource in violation. If there isn't an `urn`, a violation is associated with the stack itself.
```

An example stack validation policy is shown below:

```typescript
import { Resource } from "@pulumi/pulumi";
import * as q from "@pulumi/pulumi/queryable";
import * as aws from "@pulumi/aws";
import { PolicyPack, ReportViolation, StackValidation, StackValidationArgs, StackValidationPolicy } from "@pulumi/policy";

const s3BucketLoggingEnabled: StackValidationPolicy = {
    name: "s3-bucket-logging-enabled",
    description: "Checks whether logging is enabled for your S3 buckets.",
    enforcementLevel: "mandatory",
    validateStack: validateTypedResources(aws.s3.Bucket, (buckets, args, reportViolation) => {
        // First, save any bucket IDs that are being used as logging targets.
        const logBucketIDs: Set<string> = new Set();
        for (const bucket of buckets) {
            if (bucket.loggings) {
                for (const logging of bucket.loggings) {
                    if (logging.targetBucket) {
                        logBucketIDs.add(logging.targetBucket);
                    }
                }
            }
        }

        // Then, check the buckets for violations.
        for (const bucket of buckets) {

>             // Skip any buckets that haven't yet been provisioned (i.e. during previews).

Reviewer 2: ```suggestion
            // Skip any buckets not provisioned (i.e., during previews).
```

            if (!bucket.id) {
                continue;
            }
            // If the bucket doesn't have any loggings and the bucket itself isn't being used as
            // a log target, it's in violation of the policy.
            if (!bucket.loggings || bucket.loggings.length === 0) {
                if (!logBucketIDs.has(bucket.id)) {
                    reportViolation("Bucket logging must be defined.", bucket.urn);
                }
            }
        }
    }),
};

> // Utility method for defining a new StackValidation that will return the resources matching the provided type.

Reviewer 2: ```suggestion
// Utility method for defining a new StackValidation that returns the resources matching the provided type.
```

function validateTypedResources<TResource extends Resource>(
    resourceClass: { new(...rest: any[]): TResource },
    validate: (
        resources: q.Resolved<TResource>[],
        args: StackValidationArgs,
        reportViolation: ReportViolation) => Promise<void> | void,
): StackValidation {
    return (args: StackValidationArgs, reportViolation: ReportViolation) => {
        const isInstance = (<any>resourceClass).isInstance;
        if (!isInstance || typeof isInstance !== "function") {
            return;
        }
        const resources = args.resources
            .filter(r => isInstance({ __pulumiType: r.type }) === true)
            .map(r => ({ ...r.props, urn: r.urn } as q.Resolved<TResource>));
        validate(resources, args, reportViolation);
    };
}
```

> The above example checks whether logging is enabled on S3 buckets. If a bucket does not have a `loggings` property defined, it will be in violation of the policy unless the bucket is being used as a logging `targetBucket` of another bucket. The policy check is skipped if the bucket's `id` isn't available, meaning this policy will only run during updates when the resource `id`s are available.

Reviewer 2: ```suggestion
The example above checks whether if the S3 buckets have logging enabled. If a bucket does not have a `loggings` property defined, it violates the policy unless the bucket is a logging `targetBucket` of another bucket. The policy check skips the check if the bucket's `id` isn't available, meaning this policy only runs during updates when the resource `id`s are available.
```

> At the time of writing, there is no built-in helper for stack validation policies similar to `validateTypedResource` for resource validation policies, so we've included our own helper that can be used until a [built-in helper](https://github.com/pulumi/pulumi-policy/issues/158) is available.

Reviewer 2: ```suggestion
At the time of writing, there is no built-in helper for stack validation policies similar to `validateTypedResource` for resource validation policies, so we've included our own helper until a [built-in helper](https://github.com/pulumi/pulumi-policy/issues/158) is available.
```

## Resource Validation vs. Stack Validation

> |                                | Resource Validation                   | Stack Validation                                     |

Reviewer: :cheff-kiss-emoji: This table, wonderful. Clearly you shouldn't put the table earlier in the doc, make it POP here, later. Wonderful.

|--------------------------------|---------------------------------------|------------------------------------------------------|
| What does it check?            | Individual resources                  | All resources in the stack                           |
| When is the check performed?   | Before resources are created/modified | After all stack resources have been created/modified |

> | What information is available? | Resource _input_ properties           | Resource _output_ properties                         |

Reviewer 3: For Stack Validation both input and output are available correct? Maybe this is implicit but i dont think its super clear

Author: It's just the outputs. https://github.com/pulumi/pulumi/blob/10a960ea4b7728a21fe0b0a6aa9768994a538e92/pkg/resource/deploy/step_generator.go#L1251-L1253

Although, during previews, inputs are propagated to outputs (as of https://github.com/pulumi/pulumi/pull/3245).

I'll add an explanation of the difference between previews and updates.

> In summary:

Reviewer: "In summary" is a super-weak segue and we should try to use stronger language. How about:

With two different ways to write policies, you have more flexibility about how you want to describe and enforce your rules:

> - Use resource validation policies when you want to check individual types of resources and block them from being created or modified, checking resource _input_ properties.

Reviewer 2: ```suggestion
- Use resource validation policies when you want to check individual types of resources and block them from being created or modified, checking resource _input_ properties.
```
how does 'checking resource _input_ properties.' fit into this sentence?

> - Use stack validation policies when you want to check the stack as a whole (all resources) after they've been created or modified, checking resource _output_ properties.

Reviewer 3: Should we also include something about best practices for writing policies? Seems semi-related but maybe not?

Author: Yeah, I too was thinking about including that and a bunch of other details, but decided to just keep the post shorter rather than longer.

> - Use stack validation policies when you want to check the stack as a whole (all resources) after they've been created or modified, checking resource _output_ properties.

Reviewer 2: ```suggestion
- Use stack validation policies when you want to check the stack as a whole (all resources) after they've been created or modified, checking resource _output_ properties.
```
unsure how "checking resource _output_ properties." fits into this sentence

## Trying It Out

> In this post, we've seen the different types of CrossGuard policies that can be written. To give it a try, [download and install Pulumi]({{< relref "/docs/get-started" >}}) and [get started with policy as code]({{< relref "/docs/get-started/crossguard" >}}). For examples of policies, check out [Pulumi CrossGuard policies for AWS (AWSGuard)](https://github.com/pulumi/pulumi-policy-aws) or our [policy examples](https://github.com/pulumi/examples/tree/master/policy-packs).

Reviewer: <p>
And as always, if you'd like to share what policies you have come up with on your own you can discuss them on the [Pulumi Community Slack].

> In this post, we've seen the different types of CrossGuard policies that can be written. To give it a try, [download and install Pulumi]({{< relref "/docs/get-started" >}}) and [get started with policy as code]({{< relref "/docs/get-started/crossguard" >}}). For examples of policies, check out [Pulumi CrossGuard policies for AWS (AWSGuard)](https://github.com/pulumi/pulumi-policy-aws) or our [policy examples](https://github.com/pulumi/examples/tree/master/policy-packs).

Reviewer 3: I would split this into two paragraphs. Adding a new line after, `In this post, we've seen the different types of CrossGuard policies that can be written.`
